Frameworks/CoreText/DWriteWrapper_CTFont.mm

```diff
@@ -208,6 +208,207 @@ HRESULT _DWriteCreateFontFaceWithFontDescriptor(CTFontDescriptorRef fontDescript
+    std::vector<ComPtr<IDWriteFont>> matchingFonts;
+    matchingFonts.reserve(fonts.size());
+    CFRange range{ 0, CFStringGetLength(key) };
+    if (CFStringFindWithOptions(key, kCTFontFamilyNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByFamilyName(fonts, static_cast<CFStringRef>(value), matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByInformationalString(fonts,
+                                                            static_cast<CFStringRef>(value),
+                                                            DWRITE_INFORMATIONAL_STRING_POSTSCRIPT_NAME,
+                                                            matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontDisplayNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByInformationalString(fonts,
+                                                            static_cast<CFStringRef>(value),
+                                                            DWRITE_INFORMATIONAL_STRING_FULL_NAME,
+                                                            matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontStyleNameAttribute, range, 0, nullptr)) {
+        RETURN_IF_FAILED(__FilterFontsByStyleName(fonts, static_cast<CFStringRef>(value), matchingFonts));
+    } else if (CFStringFindWithOptions(key, kCTFontTraitsAttribute, range, 0, nullptr)) {
+        __FilterFontsByTraits(fonts, static_cast<CFDictionaryRef>(value), matchingFonts);
+    } else {
+        // Other attributes are currently unsupported by CTFont
+        UNIMPLEMENTED();
+    }
+
+    // Do not filter if there's no match and forceMatch isn't true
+    // Used to share filtering between mandatory and optional attributes
+    if (!matchingFonts.empty() || forceMatch) {
+        fonts = std::move(matchingFonts);
+    }
+
+    return S_OK;
+}
+
+#pragma endregion // Helper Filter Methods
+
+/**
+ * Creates a CFArray of CTFontDescriptors values for all fonts that match the given attributes best and completely match mandatoryKeys
+ */
+HRESULT _DWriteCreateMatchingFontDescriptors(CFDictionaryRef attributes, CFSetRef mandatoryKeys, CFArrayRef* matchingNames) {
+    RETURN_HR_IF(E_POINTER, !attributes || !matchingNames);
+    *matchingNames = nullptr;
+
+    std::vector<ComPtr<IDWriteFont>> fonts;
+    if (CFDictionaryContainsKey(attributes, kCTFontNameAttribute) &&
+        (!mandatoryKeys || CFSetContainsValue(mandatoryKeys, kCTFontNameAttribute))) {
+        // Font name is either mandatory or there are no mandatory keys, so we can start match from font
+        ComPtr<IDWriteFont> matchingFont;
+        if (FAILED(_DWriteCreateFontWithName(static_cast<CFStringRef>(CFDictionaryGetValue(attributes, kCTFontNameAttribute)),
```

Comment:
ugh you were right, this if nesting is really ugly...

I think I'd prefer the failure case second, since there's a case where it just no-ops, and it's more intuitive to see that if the eye doesn't have to read the else block afterward